staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1012 @@
+				`has(min.strVal) && min.strVal.matches(r'(\d+(\.\d+)?%)')`,
+				// validate that 'min' can be either a exact value minimum, or a minimum as a percentage of 'available'
+				"has(min.intVal) ? current <= min.intVal : double(current) / double(available) >= double(min.strVal.replace('%', '')) / 100.0",
+			},
+		},
+		{name: "unknown",
+			obj: map[string]interface{}{
+				"unknown": map[string]interface{}{
+					"field1": "a",
+					"field2": "b",
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"unknown": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XPreserveUnknownFields: true,
+					},
+				},
+			}),
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(self.field1)": "undefined field 'field1'",
+			},
+		},
+		{name: "known and unknown fields",
+			obj: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"minReplicas": 1,
+					"maxReplicas": 2,
+					"field1":      "a",
+				},
+			},
+			schema: &schema.Structural{
+				Generic: schema.Generic{
+					Type: "object",
+				},
+				Properties: map[string]schema.Structural{
+					"spec": {
+						Generic: schema.Generic{Type: "object"},
+						Extensions: schema.Extensions{
+							XPreserveUnknownFields: true,
+						},
+						Properties: map[string]schema.Structural{
+							"minReplicas": {
+								Generic: schema.Generic{
+									Type: "integer",
+								},
+							},
+							"maxReplicas": {
+								Generic: schema.Generic{
+									Type: "integer",
+								},
+							},
+						},
+					},
+				},
+			},
+			valid: []string{"spec.minReplicas < spec.maxReplicas"},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(spec.field1)": "undefined field 'field1'",
+			},
+		},
+		{name: "escaping",
+			obj: map[string]interface{}{
+				// RESERVED symbols defined in the CEL lexer
+				"true": 1,
+				// identifiers that are part of the CEL language
+				"int": 2,
+				// validation expression reserved identifiers
+				"self": 3,
+				// identifiers that are CEL builtin function and macro names
+				"getDate": 4,
+				"all":     5,
+				// identifiers that collide with escaped identifiers
+				"_true":  6,
+				"__true": 7,
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"true":    integerType,
+				"int":     integerType,
+				"self":    integerType,
+				"getDate": integerType,
+				"all":     integerType,
+				"_true":   integerType,
+				"__true":  integerType,
+			}),
+			valid: []string{
+				// CEL lexer RESERVED keywords must be escaped
+				"_true == 1",
+				"self._true == 1",
+				// CEL language identifiers do not need to be escaped, but would collide with builtin language identifier if bound as
+				// root variable names, so are only field selectable as a field of 'self'.
+				"self.int == 2",
```

Comment:
`_int == 2` should work too, right?

Reply:
Right now we're only allowing access to identifiers that conflict with language identifiers it through `self.int`.